8329332: Remove CompiledMethod and CodeBlobLayout classes

vnkozlov opened the pull request:
Revert [JDK-8152664](https://bugs.openjdk.org/browse/JDK-8152664) RFE [changes](https://github.com/openjdk/jdk/commit/b853eb7f5ca24eeeda18acbb14287f706499c365) which was used for AOT [JEP 295](https://openjdk.org/jeps/295) implementation in JDK 9. The code was left in HotSpot assuming it will help in a future. But during work on Leyden we decided to not use it. In Leyden cached compiled code will be restored in CodeCache as normal nmethods: no need to change VM's runtime and GC code to process them.

I may work on optimizing `CodeBlob` and `nmethod` fields layout to reduce header size in separate changes. In these changes I did simple fields reordering to keep small (1 byte) fields together.

I do not see (and not expected) performance difference with these changes.

Tested tier1-5, xcomp, stress. Running performance testing.

I need help with testing on platforms which Oracle does not support.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329332](https://bugs.openjdk.org/browse/JDK-8329332): Remove CompiledMethod and CodeBlobLayout classes (**Enhancement** - P4)


### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [246ff68a](https://git.openjdk.org/jdk/pull/18554/files/246ff68aab3f093b642cc09041bb939a18dc0301)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18554/head:pull/18554` \
`$ git checkout pull/18554`

Update a local copy of the PR: \
`$ git checkout pull/18554` \
`$ git pull https://git.openjdk.org/jdk.git pull/18554/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18554`

View PR using the GUI difftool: \
`$ git pr show -t 18554`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18554.diff">https://git.openjdk.org/jdk/pull/18554.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18554#issuecomment-2027868356)

RealFYang commented:
Hi, I also performed some tests (tier1-3 and hotspot:tier4) on linux-riscv64 platform. Result looks good.

offamitkumar commented:
I performed the build + testing `{fastdebug, release, slowdebug} X {tier1}`  on `s390x` and result looks fine.

dean-long commented:
The `not_used` state was introduced for AOT.  It can go away now.

vnkozlov commented:
@RealFYang and @offamitkumar  thank you for testing.

vnkozlov commented:
> The `not_used` state was introduced for AOT. It can go away now.

Good catch, Dean.

I want to keep `nmethod::make_not_used()` method because we use it in Leyden to keep AOT code (outside of CodeCache): [nmethod.hpp#L476](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/code/nmethod.hpp#L476)
It does not use this flag value.

vnkozlov commented:
I did not change `src/hotspot/share/code//codeHeapState.cpp` code which counts nmethods with `not_used` state by checking `(!nm->is_not_entrant()` after `(nm->is_in_use())`.  Removing `not_used` does not affect it.
The code is complicated and needs separate RFE if we decide to clean it.

iwanowww reviewed (approved):
Nice cleanup! Overall, looks very good.

You removed `not_used` nmethod state. What about `nmethod::make_not_used()`?

What about `CompiledMethod_lock`? There's no `CompiledMethod` anymore, but the lock name still refers to it.

stefank commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -417,6 +425,558 @@ static int adjust_pcs_size(int pcs_size) {
+    // nmethods found in safepoints have gone through an entry barrier and are not armed.
+    // By calling this nmethod entry barrier, it plays along and acts
+    // like any other nmethod found on the stack of a thread (fewer surprises).
+    nmethod* nm = as_nmethod_or_null();
```
Calling as_nmethod_or_null() from within functions in the nmethod class is suspicious. Shouldn't all such usages be removed? (I'm fine with doing that as a separate change)

stefank replied:
Sounds good.

stefank commented on src/hotspot/share/code/nmethod.cpp:
```diff
@@ -1587,7 +2161,7 @@ void nmethod::post_compiled_method_unload() {
   DTRACE_METHOD_UNLOAD_PROBE(method());
 
-  // If a JVMTI agent has enabled the CompiledMethodUnload event then
+  // If a JVMTI agent has enabled the nmethod Unload event then
```
I think the event is still called CompiledMethodUnload, so this line should probably be reverted.

stefank commented on src/hotspot/share/compiler/compileBroker.cpp:
```diff
@@ -1481,7 +1481,7 @@ nmethod* CompileBroker::compile_method(const methodHandle& method, int osr_bci,
   // We accept a higher level osr method
   if (osr_bci == InvocationEntryBci) {
-    CompiledMethod* code = method->code();
+    nmethod* code = method->code();
```
Cast below is redundant.

stefank commented on src/hotspot/share/gc/g1/g1HeapRegion.cpp:
```diff
@@ -336,7 +336,7 @@ class VerifyCodeRootCodeBlobClosure: public CodeBlobClosure {
 
   void do_code_blob(CodeBlob* cb) {
-    nmethod* nm = (cb == nullptr) ? nullptr : cb->as_compiled_method()->as_nmethod_or_null();
+    nmethod* nm = (cb == nullptr) ? nullptr : cb->as_nmethod_or_null();
```
After this change I'd like to see if we can change this and similar GC interfaces to work directly against `nmethod` instead of `CodeBlob`.

stefank commented on src/hotspot/share/gc/shared/parallelCleaning.cpp:
```diff
@@ -49,15 +49,15 @@ CodeCacheUnloadingTask::~CodeCacheUnloadingTask() {
-  CompiledMethodIterator last(CompiledMethodIterator::all_blobs);
+void CodeCacheUnloadingTask::claim_nmethods(nmethod** claimed_nmethods, int *num_claimed_nmethods) {
+  nmethod* first;
+  NMethodIterator last(NMethodIterator::all_blobs);
```
FWIW, `all_blobs` is slightly confusing name when nmethods are a subset of all "code blobs". We might want to consider renaming this to `NMethodIterator::all` (in a separate RFE).

stefank commented on src/hotspot/share/gc/x/xUnload.cpp:
```diff
@@ -75,7 +75,7 @@ class XIsUnloadingOopClosure : public OopClosure {
 class XIsUnloadingBehaviour : public IsUnloadingBehaviour {
 public:
-  virtual bool has_dead_oop(CompiledMethod* method) const {
+  virtual bool has_dead_oop(nmethod* method) const {
```
This now takes an `nmethod` argument, but still calls as_nmethod(). I think that should be removed from this, and all similar functions here in the GC code. If you want, I can do that as a follow-up RFE.

vnkozlov replied:
I decided to fix it in these changes.

stefank commented on src/hotspot/share/jvmci/jvmciRuntime.cpp:
```diff
@@ -267,10 +267,10 @@ JRT_ENTRY_NO_ASYNC(static address, exception_handler_for_pc_helper(JavaThread* c
 
-  cm = CodeCache::find_compiled(pc);
-  assert(cm != nullptr, "this is not a compiled method");
+  nm = CodeCache::find_nmethod(pc);
+  assert(nm != nullptr, "this is not a compiled method");
```
Unrelated to this patch, but might be worth mentioning because I think it would be good to think about this in a follow-up RFE. `CodeCache::find_blob` returns null if it can't find a matching blob, but `CodeCache::find_nmethod` asserts that it did find one. The latter makes the assert redundant, but it also begs to question if `find_blob` and `find_nmethod` realy should be different in this regard? Should we have `find_blob_or_null` and `find_nmethod_or_null`? Alt. `find_blob_not_null` and `find_nmethod_not_null`.

stefank commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -768,9 +768,9 @@ class VM_WhiteBoxDeoptimizeFrames : public VM_WhiteBoxOperation {
-                assert(cm != nullptr, "sanity check");
-                cm->make_not_entrant();
+                nmethod* nm = CodeCache::find_nmethod(f->pc());
+                assert(nm != nullptr, "sanity check");
```
This assert is now redundant.

stefank commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -1097,7 +1097,7 @@ bool WhiteBox::compile_method(Method* method, int comp_level, int bci, JavaThrea
   // Check code again because compilation may be finished before Compile_lock is acquired.
   if (bci == InvocationEntryBci) {
-    CompiledMethod* code = mh->code();
+    nmethod* code = mh->code();
```
`as_nmethod_or_null()` below should be redundant.

stefank commented on src/hotspot/share/runtime/continuationEntry.hpp:
```diff
@@ -32,7 +32,7 @@
 #include CPU_HEADER(continuationEntry)
 
-class CompiledMethod;
+class nmethod;
```
Maybe keep the forward declarations sorted?

stefank commented on src/hotspot/share/runtime/continuationEntry.hpp:
```diff
@@ -56,12 +56,12 @@ class ContinuationEntry {
 public:
   static int _return_pc_offset; // friend gen_continuation_enter
-  static void set_enter_code(CompiledMethod* cm, int interpreted_entry_offset);
+  static void set_enter_code(nmethod* cm, int interpreted_entry_offset);
```
cm => nm?

stefank commented on src/hotspot/share/runtime/frame.cpp:
```diff
@@ -205,11 +205,11 @@ void RegisterMap::print() const {
-    CompiledMethod* cm = cb()->as_compiled_method_or_null();
-    if (cm->is_method_handle_return(pc()))
-      return cm->deopt_mh_handler_begin() - pc_return_offset;
+    nmethod* nm = cb()->as_nmethod_or_null();
```
Prexisting: It's weird that this code is using the `_or_null()` version when the code below does not null check the returned value.

vnkozlov replied:
Before [JDK-6921352](https://bugs.openjdk.org/browse/JDK-6921352) it was:
```
  return ((nmethod*) cb())->deopt_handler_begin() - pc_return_offset;
```

I will add assert with check for null. We definitely expect here only nmethod.

stefank commented on src/hotspot/share/runtime/vframe.cpp:
```diff
@@ -71,8 +71,8 @@ vframe* vframe::new_vframe(const frame* f, const RegisterMap* reg_map, JavaThrea
-    if (cb->is_compiled()) {
-      CompiledMethod* nm = (CompiledMethod*)cb;
+    if (cb->is_nmethod()) {
+      nmethod* nm = cb->as_nmethod();;
```
There's two `;`s on this line.

stefank reviewed (commented):
Nice!

We've wanted to clean up some interfaces between the CodeCache and the GC code by using nmethod closures instead of CodeBlob closures. This change (and the Sweeper removal) makes it possible to do those cleanups.

I've made a superficial pass over the patch to and left a few comments. Most of those comments are things that would be nice to fix, but could also be left as follow-up RFEs (if they are deemed to be worthy ideas to pursue).

vnkozlov commented:
> What about `CompiledMethod_lock`? There's no `CompiledMethod` anymore, but the lock name still refers to it.

It was different changes [JDK-8226705](https://bugs.openjdk.org/browse/JDK-8226705). Renaming it will complicate these changes more than I wanted. I can do it in separate RFE.

stefank commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -506,17 +403,14 @@ class RuntimeStub: public RuntimeBlob {
 
   // GC/Verification support
-  void preserve_callee_argument_oops(frame fr, const RegisterMap *reg_map, OopClosure* f)  { /* nothing to do */ }
+  virtual void preserve_callee_argument_oops(frame fr, const RegisterMap *reg_map, OopClosure* f) override { /* nothing to do */ }
```
`override` is not required by C++. You do however mark all virtual methods with `override` if any of the functions are marked with `override`. I think it would be good to have a HotSpot code style discussion about this (but not in this RFE).

vnkozlov replied:
I put `virtual/override` cleanup in CodeBlob as additional suggestion in followup RFE JDK-8329628.

stefank commented on src/hotspot/share/code/codeCache.cpp:
```diff
@@ -1011,7 +1006,7 @@ int CodeCache::nmethod_count(CodeBlobType code_blob_type) {
 int CodeCache::nmethod_count() {
   int count = 0;
-  FOR_ALL_NMETHOD_HEAPS(heap) {
+  for (GrowableArrayIterator<CodeHeap*> heap = _nmethod_heaps->begin(); heap != _nmethod_heaps->end(); ++heap) {
```
I didn't say that blob_count used `FOR_ALL_NMETHODS_HEAP`. I wrote "one of these macros". I still think this adds an inconsistency to the code that I don't think is beneficial.

With that said, can't this be written as:
```
for (CodeHeap* heap : *_nmethod_heaps) {
```

Maybe yet another opportunity for cleanups.

vnkozlov replied:
I like it and will do it in JDK-8329628.

vnkozlov commented:
Thank you, @stefank, for great review. I addressed all your comments locally and with run testing in mach5 before pushing it.

Except your suggestion about `find_blob_not_null()` - should be separate RFE.
The same for suggestion "GC interfaces to work directly against nmethod instead of CodeBlob".

vnkozlov commented:
I filed RFEs:
[JDK-8329628](https://bugs.openjdk.org/browse/JDK-8329628): Additional changes after JDK-8329332
[JDK-8329629](https://bugs.openjdk.org/browse/JDK-8329629): GC interfaces should work directly against nmethod instead of CodeBlob

vnkozlov commented:
GHA `linux-x64-hs-minimal` failure is not related to changes:
```
2024-04-04T00:07:46.9654262Z ##[warning]Failed to download action 'https://api.github.com/repos/actions/github-script/tarball/60a0d83039c74a4aee543508d2ffcb1c3799cdea'. Error: The request was canceled due to the configured HttpClient.Timeout of 100 seconds elapsing.
2024-04-04T00:07:46.9656929Z ##[warning]Back off 22.252 seconds before retry.
2024-04-04T00:08:52.1252710Z ##[error]The SSL connection could not be established, see inner exception.
```

xmas92 reviewed (commented):
There is a stale comment in `test/jdk/com/sun/jdi/EATests.java:1288`

```diff
-// (See CompiledMethod::is_at_poll_return())
+// (See nmethod::is_at_poll_return())
```

stefank commented on src/hotspot/share/gc/shared/gcBehaviours.hpp:
```diff
@@ -28,14 +28,14 @@
 #include "oops/oopsHierarchy.hpp"
 
-// This is the behaviour for checking if a CompiledMethod is unloading
+// This is the behaviour for checking if a nmethod is unloading
```
Maybe this should be *an* nmethod?

vnkozlov replied:
Quote: "an" goes before words that begin with vowels.

stefank replied:
I don't think that holds if the 'n' is pronounced the way nmethod is pronounced.

vnkozlov replied:
`grep` shows that we have both cases but `an nmethod` is used more.

vnkozlov replied:
I will fix it here as you suggested but I am not touching other places.

stefank commented on src/hotspot/share/gc/shenandoah/shenandoahUnload.cpp:
```diff
@@ -77,8 +78,7 @@ class ShenandoahIsUnloadingOopClosure : public OopClosure {
 public:
-  virtual bool has_dead_oop(CompiledMethod* method) const {
-    nmethod* const nm = method->as_nmethod();
+  virtual bool has_dead_oop(nmethod* const nm) const {
```
Is there a reason why this was changed to `nmethod* const nm` instead of `nmethod* nm`? IsUnloadingBehviour::has_dead_oop uses `nmethod* nm`. This question applies to the other changes in this file as well.

stefank commented on src/hotspot/share/gc/x/xUnload.cpp:
```diff
@@ -75,8 +75,7 @@ class XIsUnloadingOopClosure : public OopClosure {
 public:
-  virtual bool has_dead_oop(CompiledMethod* method) const {
-    nmethod* const nm = method->as_nmethod();
+  virtual bool has_dead_oop(nmethod* const nm) const {
```
`nmethod* const nm` => `nmethod* nm`. (ZGC's style is to use const for local variables, but not for variables in the parameter list). The same applies to the rest of the changes to this file.

vnkozlov replied:
Okay. I did not know that it is only used for locals. I will update code.

stefank commented on src/hotspot/share/gc/z/zUnload.cpp:
```diff
@@ -74,8 +74,7 @@ class ZIsUnloadingOopClosure : public OopClosure {
 public:
-  virtual bool has_dead_oop(CompiledMethod* method) const {
-    nmethod* const nm = method->as_nmethod();
+  virtual bool has_dead_oop(nmethod* const nm) const {
```
`nmethod* const nm` => `nmethod* nm`. (ZGC's style is to use const for local variables, but not for variables in the parameter list). The same applies to the rest of the changes to this file.

stefank commented on src/hotspot/share/runtime/javaThread.hpp:
```diff
@@ -120,7 +120,7 @@ class JavaThread: public Thread {
   DeoptResourceMark*  _deopt_mark;               // Holds special ResourceMark for deoptimization
 
-  CompiledMethod*       _deopt_nmethod;         // CompiledMethod that is currently being deoptimized
+  nmethod*       _deopt_nmethod;                 // nmethod that is currently being deoptimized
```
The alignment is (and was) weird here.

stefank commented on src/hotspot/share/code/codeBlob.hpp:
```diff
@@ -155,17 +155,17 @@ class CodeBlob {
-  bool is_vtable_blob() const                 { return _kind == Blob_Vtable; }
-  bool is_method_handles_adapter_blob() const { return _kind == Blob_MH_Adapter; }
-  bool is_upcall_stub() const                 { return _kind == Blob_Upcall; }
+  bool is_nmethod() const                     { return _kind == CodeBlobKind::Blob_Nmethod; }
+  bool is_buffer_blob() const                 { return _kind == CodeBlobKind::Blob_Buffer; }
+  bool is_runtime_stub() const                { return _kind == CodeBlobKind::Blob_Runtime_Stub; }
+  bool is_deoptimization_stub() const         { return _kind == CodeBlobKind::Blob_Deoptimization; }
+  bool is_uncommon_trap_stub() const          { return _kind == CodeBlobKind::Blob_Uncommon_Trap; }
+  bool is_exception_stub() const              { return _kind == CodeBlobKind::Blob_Exception; }
+  bool is_safepoint_stub() const              { return _kind == CodeBlobKind::Blob_Safepoint; }
+  bool is_adapter_blob() const                { return _kind == CodeBlobKind::Blob_Adapter; }
+  bool is_vtable_blob() const                 { return _kind == CodeBlobKind::Blob_Vtable; }
+  bool is_method_handles_adapter_blob() const { return _kind == CodeBlobKind::Blob_MH_Adapter; }
+  bool is_upcall_stub() const                 { return _kind == CodeBlobKind::Blob_Upcall; }
```
The `Blob_` prefix is now redundant since we always have to prefix with CodeBlobKind::. Just a suggestion if you want to shorten these.

vnkozlov replied:
Good suggestion

stefank reviewed (approved):
I took a second pass over the changes. I've given a few suggestions below. None of them should require respinning of tests (except for making sure that this still builds).

stefank reviewed (approved):
Looks good.

vnkozlov commented:
Thank you, @stefank, @iwanowww, @dean-long  and @xmas92 for reviews and @RealFYang and @offamitkumar  for testing.

vnkozlov commented:
/integrate
